mapper/orm/src/main/java/org/hibernate/search/mapper/orm/event/impl/HibernateSearchEventListener.java

```diff
@@ -131,16 +133,32 @@ public void onPostUpdateCollection(PostCollectionUpdateEvent event) {
 	 */
 	@Override
 	public void onFlush(FlushEvent event) {
-		// TODO HSEARCH-3068 handle the "simulated" transaction when a Flush listener is registered
-//		Session session = event.getSession();
-//		Synchronization synchronization = flushSynch.get( session );
-//		if ( synchronization != null ) {
-//			//first cleanup
-//			flushSynch.remove( session );
-//			log.debug( "flush event causing index update out of transaction" );
-//			synchronization.beforeCompletion();
-//			synchronization.afterCompletion( Status.STATUS_COMMITTED );
-//		}
+		HibernateOrmListenerContextProvider contextProvider = state.getContextProvider();
+		EventSource session = event.getSession();
+		PojoWorkPlan currentWorkPlan = getCurrentWorkPlan( contextProvider, session );
+		currentWorkPlan.prepare();
+
+		// flush within a transaction should trigger only the prepare phase,
+		// since the execute phase is supported to be triggered by the transaction commit
+		if ( !session.isTransactionInProgress() ) {
+			// out of transaction it will trigger both of them
+			AutomaticIndexingSynchronizationStrategy synchronizationStrategy = contextProvider.getSynchronizationStrategy();
+			synchronizationStrategy.handleFuture( currentWorkPlan.execute() );
+		}
+	}
+
+	@Override
+	public void onAutoFlush(AutoFlushEvent event) throws HibernateException {
+		getCurrentWorkPlan( state.getContextProvider(), event.getSession() ).prepare();
+	}
+
+	@Override
+	public void onClear(ClearEvent event) {
+		EventSource session = event.getSession();
+		// skip the clearNotPrepared operation in case of rollback
+		if ( session.isTransactionInProgress() ) {
+			getCurrentWorkPlan( state.getContextProvider(), session ).clearNotPrepared();
```

Comment:
I get why you originally added this, but I think it's wrong now that you support flush events outside of transactions. What happens when you change entities, clear, then flush, outside of a transaction? It shoudn't do anything, yet with this code I suspect it will update the index, or result in a LazyInitializationException, because the clear is ignored.

What was the problem with executing this code outside of a transaction, exactly?

Reply:
We would have a small issue just on rollback. Since the rollback triggers a clear event out of transaction. The issue is just about the creation of a work plan for nothing,
But we can change the predicate into:
``` java
// skip the clearNotPrepared operation in case of rollback
if ( !TransactionStatus.ROLLED_BACK.equals( session.getTransaction().getStatus() ) ) {
  getCurrentWorkPlan( state.getContextProvider(), session ).clearNotPrepared();
}
```

And I have a **change entities, clear, then flush, outside of a transaction** test to add